cmd/derper/cert.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"crypto/tls"
+	"fmt"
+	"log"
+	"net/http"
+	"path/filepath"
+	"regexp"
+
+	"golang.org/x/crypto/acme/autocert"
+)
+
+var unsafeHostnameCharacters = regexp.MustCompile(`[^a-zA-Z0-9-\.]`)
+
+type certProvider interface {
+	// TLSConfig provide TLS related stuff for HTTPS library
+	TLSConfig() *tls.Config
+	// HTTPHandler handle ACME related request, if any
+	HTTPHandler(fallback http.Handler) http.Handler
+}
+
+func isAvailableCertMode(mode string) bool {
+	switch mode {
+	case "letsencrypt", "manual":
+		return true
+	}
+	return false
+}
+
+func certProviderByCertMode(mode string, dir string) certProvider {
+	if dir == "" {
+		log.Fatalf("missing cert directory")
+	}
+	switch mode {
+	case "letsencrypt":
+		certManager := &autocert.Manager{
+			Prompt:     autocert.AcceptTOS,
+			HostPolicy: autocert.HostWhitelist(*hostname),
+			Cache:      autocert.DirCache(*certDir),
+		}
+		if *hostname == "derp.tailscale.com" {
+			certManager.HostPolicy = prodAutocertHostPolicy
+			certManager.Email = "security@tailscale.com"
+		}
+		return certManager
+	case "manual":
```

Comment:
kinda a bummer to duplicate this switch/check between the two funcs. Maybe this can return (certProvider, error) instead and then you can remove `func isAvailableCertMode`?